auto domain adjustment

psychocoderHPC opened the pull request:
Add helper to calculate/adjust the domain sizes and offset of the simulation.

- add class `AdjustDomainSize`
- remove domain validation from `MySimulation`
- add new command line parameter `--autoAdjustGrid`


If the domain is automatically adjusted the changes will be logged on `stdout`. I know that this is not optimal but since we are lacking a logger this is the best way.

sbastrakov commented on include/picongpu/simulationControl/AdjustDomainSize.hpp:
```diff
@@ -0,0 +1,363 @@
+        ) :
+            m_numDevices( numDevices),
+            m_mpiPosition( mpiPosition ),
+            m_isPeriodic( isPeriodic),
```
`( isPeriodic)` -> `( isPeriodic )`

sbastrakov commented on include/picongpu/simulationControl/AdjustDomainSize.hpp:
```diff
@@ -0,0 +1,363 @@
+        {
+        }
+
+        /** adjust the correct grid size
```
To adjust already implies to correct, so `correct` here is excess.

sbastrakov commented on include/picongpu/simulationControl/AdjustDomainSize.hpp:
```diff
@@ -0,0 +1,363 @@
+                updateLocalDomainOffset( d );
+            }
+
+            // write results back
```
Question: is it not possible have smth like `sizeChanged = (m_globalDomainSize != m_globalDomainSize) || (m_localDomainSize !=  localDomainSize) || (m_localDomainOffset != localDomainOffset )` here? (Instead of having it in different places throughout this class)

sbastrakov commented on include/picongpu/simulationControl/AdjustDomainSize.hpp:
```diff
@@ -0,0 +1,363 @@
+         *
+         * @param dim dimension to update
+         */
+        void minThreeSuperCells( size_t dim )
```
Not important, but in other similar situations `dim` is `const`.

sbastrakov commented on include/picongpu/simulationControl/AdjustDomainSize.hpp:
```diff
@@ -0,0 +1,363 @@
+        {
+            int superCellCount = m_localDomainSize[ dim ] / SuperCellSize::toRT()[ dim ];
+            if( superCellCount < 3 )
+                superCellCount = 3;
```
Maybe use `std::max` instead of this if statement?

sbastrakov commented on include/picongpu/simulationControl/AdjustDomainSize.hpp:
```diff
@@ -0,0 +1,363 @@
+            int validLocalSize = m_localDomainSize[ dim ];
+            if(
+                !m_isPeriodic[ dim ] &&
+                // device has absorber
```
Maybe a self-explanatory variables would be better than the comment? Like 
```
bool const absorberEnabled = !m_isPeriodic[ dim ];
bool const isBoundaryDevice = ( m_mpiPosition[ dim ] == 0 || m_mpiPosition[ dim ] == m_numDevices[ dim ] - 1 );
bool const hasAbsorber = absorberEnabled && isBoundaryDevice ;
if (hasAbsorber) ...
```

sbastrakov commented on include/picongpu/simulationControl/AdjustDomainSize.hpp:
```diff
@@ -0,0 +1,363 @@
+                ( m_mpiPosition[ dim ] == 0 || m_mpiPosition[ dim ] == m_numDevices[ dim ] - 1 )
+            )
+            {
+                int const maxAbsorberCells = static_cast< int >(
```
I might have misunderstood it, but the logic of taking the maximum seems wrong to me. What if we have multiple devices along the current direction and the right-side absorber is longer (more cells) than left-side. Then it is possible that the left-most local domain is large enough to fit the left absorber, but not enough for the right absorber, and the right-most local domain is large enough. As I understand it, this case is perfectly fine, but the check would fail.

psychocoderHPC replied:
That is correct I will update this part and handle the cases correct.

sbastrakov commented on include/picongpu/simulationControl/AdjustDomainSize.hpp:
```diff
@@ -0,0 +1,363 @@
+        /** derive the local domain size
+         *
+         * Calculate the local domain size.
+         * This function take into account that the local domain size must be
```
`take` -> `takes` (or remove `This function` and start with `Take`).

sbastrakov commented on include/picongpu/simulationControl/AdjustDomainSize.hpp:
```diff
@@ -0,0 +1,363 @@
+                    &globalMax,
+                    &m_localDomainSize[ dim ],
+                    1,
+                    pmacc::mpi::reduceMethods::AllReduce()
```
Out of curiosity, why not use the version of this method without the last parameter?

sbastrakov replied:
I'm not sure either, hence asked.

sbastrakov commented on include/picongpu/simulationControl/AdjustDomainSize.hpp:
```diff
@@ -0,0 +1,363 @@
+            int validLocalSize = superCellCount * SuperCellSize::toRT()[ dim ];
+            if( validLocalSize != m_localDomainSize[ dim ] )
+            {
+                std::cout << "Dimension " << dim <<
```
I guess for a user it is more natural to operate and see log messages in terms of x, y, z, not "dimension 0 / 1 / 2". Also in case, e.g. the output only has "dimension 1", without looking to the code it is not even clear if it means x or y.

sbastrakov commented on include/picongpu/simulationControl/MySimulation.hpp:
```diff
@@ -213,16 +214,26 @@ class MySimulation : public SimulationHelper<simDim>
         {
-            gridSizeLocal[dim] = global_grid_size[dim] / gpus[dim];
-            gridOffset[dim] = gridSizeLocal[dim] * myGPUpos[dim];
+            gridSizeLocal[dim] = globalGridSize[dim] / gpus[dim];
```
In case it does not lead to too much changes, maybe rename one of `gridSizeLocal` and `globalGridSize` to make the names uniform?

sbastrakov commented on include/picongpu/simulationControl/MySimulation.hpp:
```diff
@@ -213,16 +214,26 @@ class MySimulation : public SimulationHelper<simDim>
+
+        DataSpace<simDim> gridOffset;
+
+        // adjust domain offset and sizes
```
I think the next line very clearly explains itself and so the comment is not needed.

ax3l commented:
Cool, now we just need a way to express relative positioning for `density.param` so targets do not hop out of the focal stop transversely (or alternatively: absolute positioning for the laser x/z focus position).

psychocoderHPC commented:
@sbastrakov Thanks for the feedback I added/refactored all

@ax3l 
>  ... (or alternatively: absolute positioning for the laser x/z focus position) ...

I think a refactoring of the lasers is the best solution. Since we have free functors for the density the user can decide if he prefers to describe the density relative or absolute.

ax3l commented:
> I think a refactoring of the lasers is the best solution.

Agreed!

psychocoderHPC commented on include/picongpu/simulationControl/DomainAdjuster.hpp:
```diff
@@ -0,0 +1,372 @@
+         * \warning `= { { ... } }` is not required by the c++11 standard but
+         * is necessary for g++ 4.9
+         */
+        std::array< char, 3 > dimNames = { { 'x', 'y', 'z' } };
```
NOTE: g++ 4.9 is not standard conform and requires extra brackets around the initializer list.

ax3l replied:
I think that's actually a C++11 thing and was lifted in C++14, as I read the other day on SO.

psychocoderHPC replied:
NO as I know the double brackets was only needed if you have not the assign operator but since gcc 4.9 was early stage of C++11 it is maybe a bug but not really important.

ax3l commented on include/picongpu/simulationControl/DomainAdjuster.hpp:
```diff
@@ -0,0 +1,372 @@
+            bool const isBoundaryDevice = ( m_mpiPosition[ dim ] == 0 || m_mpiPosition[ dim ] == m_numDevices[ dim ] - 1 );
+            if( isAbsorberEnabled && isBoundaryDevice )
+            {
+                size_t boundary = m_mpiPosition[ dim ] == 0u ? 0 : 1u;
```
`0u: 1u;`

ax3l commented on include/picongpu/simulationControl/DomainAdjuster.hpp:
```diff
@@ -0,0 +1,372 @@
+                size_t boundary = m_mpiPosition[ dim ] == 0u ? 0 : 1u;
+                int maxAbsorberCells = ABSORBER_CELLS[ dim ][ boundary ];
+
+                if( m_movingWindowEnabled && dim == 1 )
```
1u

ax3l commented on include/picongpu/simulationControl/DomainAdjuster.hpp:
```diff
@@ -0,0 +1,372 @@
+         */
+        void deriveLocalDomainSize( size_t const dim )
+        {
+            if( m_movingWindowEnabled && dim == 1 )
```
1u

ax3l commented on include/picongpu/simulationControl/DomainAdjuster.hpp:
```diff
@@ -0,0 +1,372 @@
+         */
+        void deriveGlobalDomainSize( size_t const dim )
+        {
+            uint64_t validGlobalGridSize = 0;
```
0u

ax3l commented on include/picongpu/simulationControl/DomainAdjuster.hpp:
```diff
@@ -0,0 +1,372 @@
+
+            deriveLocalDomainSize( dim );
+
+            if( m_movingWindowEnabled && dim == 1 )
```
1u

psychocoderHPC commented on include/picongpu/simulationControl/DomainAdjuster.hpp:
```diff
@@ -0,0 +1,372 @@
+            );
+
+            int validLocalSize = superCellCount * SuperCellSize::toRT()[ dim ];
+            if( validLocalSize != m_localDomainSize[ dim ] )
```
I would like it as it is because all rules are written in the same way. 

```
- calculate size based on the condition
- check if conditional size differ to old size
- set correct size
```

ax3l replied:
ok, but the check here can be a three-liner and one might not want to outsmart the reader ;-)

psychocoderHPC replied:
I rewrote this part.

psychocoderHPC commented on include/picongpu/simulationControl/MySimulation.hpp:
```diff
@@ -213,16 +214,24 @@ class MySimulation : public SimulationHelper<simDim>
-        Environment<simDim>::get().initGrids(global_grid_size, gridSizeLocal, gridOffset);
+        DataSpace<simDim> gridOffset;
+
+        DomainAdjuster domainAdjuster(
```
good idea. But I would say it must be enabled by default.

ax3l replied:
ok with me, so a `--no-resize-domain` flag then?

psychocoderHPC commented:
@ax3l I added the parameter `autoAdjustGrid` and updated the docs.
The reason why this parameter is called `autoAdjustGrid` and not `autoAdjustDomain` is that the other option for the user to set the simulation volume size is also called `-g,--grid`

ax3l commented on docs/source/usage/workflows/numberOfCells.rst:
```diff
@@ -8,10 +8,14 @@ Setting the Number of Cells
+#. for non periodic boundaries the absorber must fit into the local volume
+
+The grid size will be automatically adjusted if the conditions above are not fulfilled. 
+This behavior can be disabled by using the command line option `--autoAdjustGrid off`
```
needs two `` in `r.st` :)

psychocoderHPC replied:
:-( ok will fix it

ax3l replied:
yeah, it's super confusing that rst and md have this subtle difference xD

ax3l commented on docs/source/usage/workflows/numberOfCells.rst:
```diff
@@ -8,10 +8,14 @@ Setting the Number of Cells
 
 #. contain an integer *multiple* of supercells
 #. at least *three* supercells
+#. for non periodic boundaries the absorber must fit into the local volume
```
I think the term `the absorber` is not something a user knows (but our slang).
maybe better write
```suggestion
#. for non periodic boundary conditions, the number of absorbing boundary cells for devices at the simulation boundary (see :ref:`grid.param <usage-params-core>`) must fit into the local volume
```

psychocoderHPC commented:
@ax3l fixed, thx for the feedback

ax3l commented:
@psychocoderHPC Just to make sure: this works well with `--gridDist`?

psychocoderHPC commented:
yes `--gridDist` is working very well too
